### DISCUSSION TITLE
adding api service

### DISCUSSION THREAD

**Discussion by ajnaf** [ajnaf/flutter_course#7] [General]
How to add api service to flutter project

**Comment by ajnaf** [comment]
اول خطوة نقوم بها هي اضافة ال
api_service.dart 
داخل ال
services folder

**Reply by nor-nor** [reply]
السلام عليكم هدا الملف موجود
داخل المجلد
موجودات كل الدوال get ,put ,post
الكود متاعهم مش مفهوم بكل

![image](https://user-images.githubusercontent.com/80286874/111916093-f9207400-8a81-11eb-821d-eb2ed20a2178.png)

**Reply by ajnaf** [reply]
توا انديرلكم ليها شرح مفصل

**Comment by ajnaf** [comment]
داخل هذا الملف نقوم بتعريف الدوال الرئيسية الخاصة ب api requests
وهي 
get, post, put, delete
هذه الدوال عادة لا تتغير ويمكن اخذها من هذا المشروع

**Reply by nor-nor** [reply]
الملف هدا مش موجود بنضيفه انا وولا كيف
؟

**Reply by ajnaf** [reply]
موجود ايه لكن انا شرحت لو انك بدير مشروع من الصفر

**Comment by ajnaf** [comment]
نقوم باخذ ال
api end point الذي نريد ان نستخدمه
ونقوم بوضعه في ال
browser
او برنامج الpostman 
لنتائج افضل
بالنسبة للbrowser 
يمكننا فقط استعمال ال
get request 
وفي حالة لا يحتاج ال 
authentication

**Comment by ajnaf** [comment]
قم باخذ json 
وضعه في هذا الموقع 
https://javiercbk.github.io/json_to_dart/
وذلك ليقوم بتحويله الى 
class 
بلغة dart 
نستطيع استعماله في المشروع

**Reply by nor-nor** [reply]
استاد احمد شوف 
class كيف طلع
![image](https://user-images.githubusercontent.com/80286874/111916489-d727f100-8a83-11eb-9482-6a075fd37e4a.png)

**Reply by nor-nor** [reply]
حطيت  ال 
class كله زي ماطلع 
فى ملف accounmodel

**Reply by ajnaf** [reply]
كويس 
فوق كلمة generate dart 
في حقل تكتب فيه اسم المودل متاعك
مثلا City
Account,
classمع مراعات اول حرف يكون كبير لان هذا بيكون اسم ال

**Reply by nor-nor** [reply]
اوكى درتها الخطوه متاع الclass 
sldji Tag
![image](https://user-images.githubusercontent.com/80286874/111919272-e1e98280-8a91-11eb-9beb-157972136266.png)

**Comment by ajnaf** [comment]
يمكن ان يكون الناتج عبارة عن 
class واحد 
او عدة classes
نقوم بوضع كل 
class
في ملف 
dart
خاص به في models folder
في هذه الحالة قمنا بوضع 
city , area , zone

**Comment by ajnaf** [comment]
في حالة ال
get request
ويكون الناتج هو قائمة 
نقوم باضافة الدالة الاتية 
وتغيير حسب نوع الترجيع 
هذا المثال لجلب قائمة من المدن
Future<List<City>> getCities() async {
    final response = await _get('cities', isAuthenticated: false);
    List<City> results = List<City>.empty(growable: true);
    if (response != null) {
      try {
        response.forEach((city) {
          results.add(City.fromJson(city));
        });
      } catch (e) {
        print(e.message);
      }
    }
    return results;
  }

**Reply by nor-nor** [reply]
وين انحطها الكود هدا في accountviewmodel?

**Reply by ajnaf** [reply]
api.dartداخل ال

**Comment by ajnaf** [comment]
شرح الدالة
//Future يعني ان الدالة تقوم بارجاع متغير يحتاج الى وقت ليتم جلبه 
//Future <here> نوع المتغير موجود داخل الاقواس 
//اي دالة future 
//تحتاج الى async 
//لنستطيع استعمال await 
// هذا الفيديو يشرح العملية بشكل واضح
https://www.youtube.com/watch?v=vCLS3GvYMb4&ab_channel=Code2Start
Future<List<City>> getCities() async {
//response  داخل   api request نقوم  بتخزين ناتج  
//json عبارة عن  responseهذا ال 
    final response = await _get('cities', isAuthenticated: false);
    List<City> results = List<City>.empty(growable: true);
// ليس فارغ response نتحقق ان ال
    if (response != null) {
      try {
// من نوع الذي نريد ترجيعه ونقوم باضافته في القائمةobjects نقوم بتكوين  
        response.forEach((city) {
          results.add(City.fromJson(city));
        });
      } catch (e) {
        print(e.message);
      }
    }

//نقوم بترجيع القائمة
    return results;
  }

**Comment by ajnaf** [comment]
كيف نقوم باستعمال الدالة وعرض البيانات داخل الview?
اولا نقوم بتغيير BaseViewModel. الى FutureViewModel<نوع البيانات التي تريد ارجاعها>

**Comment by ajnaf** [comment]
سيظهر لك خطا تحت اسم viewmdel class 
هذا الخطا يطلب منك اضافة الدالة المطلوبة 
نقوم بعمل ctrl+. ونعمل اضافة للmissing method

**Comment by ajnaf** [comment]
اولا نقوم بتعرف object من الapiService 
final _api = locator<ApiService>();

**Comment by ajnaf** [comment]
ثم نقوم بترجيع الدالة داخل دالة futureToRun
ميزة هادي الدالة 
تخبرنا ما اذا كان الviewModel مشغول ام لا
اي ينتظر وصول البيانات من الapi
وعند جلب البيانات يضعها في متغير اسمه data لنستطيع الوصول اليها داخل الview

**Comment by ajnaf** [comment]
بالنسبة للview 
اولا يجب ان نتحقق ما اذا كان الviewModel مشغول ام لا اذا كان مشغول نقوم بعرض loading 
نستطيع معرفة اذا كان مشغول ام لا عن طريق المتغير isBusy وهو موجود داخل الFutureViewModel وميزة دالة futureToRun انه اذا كان مازال ينتظر ال api يقوم بتغير هذه القيمة لtrue وعند اتمام العملية يقوم بتغيرها تلى false مع استدعاء دالة notifyListeners() التي تقوم باعلام الواجهة ان قمنا بتغير ما ويجب ان تقومي بهذا التغيير

**Comment by ajnaf** [comment]
بالنسبة لعرض البيانات في الview
نستطيع الوصول للبيانات عن طريق model.data
وهي ما تقوم بترجيعه دالة futureToRun

ونقوم بعرضها داخل listView
